### PR TITLE
Added support for custom http.Agent. Fixes #91

### LINKED ISSUE #91: HTTP proxy support?
Hey guys,

Sorry if this was asked before, but does this client (and more generally the NodeJS APM client) support a HTTP(s) proxy? I've quickly scrolled through the docs and I couldn't find a way to configure my proxy... Is there some undocumented, automagic environment variable ?

Regards,
David

### PR DESCRIPTION
**Pull request by ddolcimascolo**
Hi guys!

Following the discussion in elastic/apm-nodejs-http-client#91 I added support for passing in custom http.Agent instances, in order to be able to traverse HTTP proxies when the APM server you're connecting to is after such a proxy.
My ultimate goal is to make https://github.com/elastic/apm-agent-nodejs support the new setting, any advice on whether I should contribute something there also is welcome :+1: 

Thanks,
David

### CODE REVIEW

**Comment by astorm** [comment]
@ddolcimascolo Hello!  Thanks for the PR, this looks like it would be a welcome addition to the agent.  

We notice that the TravisCI tests [are failing here](https://travis-ci.org/github/elastic/apm-nodejs-http-client/builds/743509682) with these changes. 

We're seeing similar failures when we run `npm run test` locally.

Do you have any thoughts on why they might be failing and what you'd need to do to fix 'em?

**Comment by ddolcimascolo** [comment]
Hi @astorm 

Thx for taking time to check this PR! I'll look into the tests failures asap and post back here. I moved a bit of configuration code in order to be able to init the custom Agent correctly, I might have overlooked something.

David

**Comment by ddolcimascolo** [comment]
@astorm I believe tests pass now :)

**Comment by astorm** [comment]
jenkins run the tests please

**Comment by astorm** [comment]
jenkins run the tests please

**Comment by ddolcimascolo** [comment]
Hi guys, I just rebased on latest master and I believe tests are passing for quite a few weeks now. Anything else you need from me before this can be merged? For the record, my goal is to make the actual nodejs APM agent compatible with this, if you want me to make a pull request there, just ask :+1: 

Cheers,
David

**Comment by astorm** [comment]
jenkins run the tests

**Comment by astorm** [comment]
@ddolcimascolo Thanks for checking in and sorry for losing you in the noise -- the squeaky wheel definitely gets the grease from a busy maintainer team :) 

I just checked out your branch and attempted to run the tests.  I saw two failures -- and that seems to line up with what our automated GitHub CI is seeing as well.


    $ npm run test

    //...

    not ok 1226 plan != count
      ---
        operator: fail
        expected: 59
        actual:   50
        at: Server.<anonymous> (/Users/astorm/Documents/github/elastic/apm-nodejs-http-client/test/central-config.js:87:11)
        stack: |-
          Error: plan != count
              at Test.assert [as _assert] (/Users/astorm/Documents/github/elastic/apm-nodejs-http-client/node_modules/tape/lib/test.js:228:54)
              at Test.bound [as _assert] (/Users/astorm/Documents/github/elastic/apm-nodejs-http-client/node_modules/tape/lib/test.js:80:32)
              at Test.fail (/Users/astorm/Documents/github/elastic/apm-nodejs-http-client/node_modules/tape/lib/test.js:322:10)
              at Test.bound [as fail] (/Users/astorm/Documents/github/elastic/apm-nodejs-http-client/node_modules/tape/lib/test.js:80:32)
              at Test._end (/Users/astorm/Documents/github/elastic/apm-nodejs-http-client/node_modules/tape/lib/test.js:173:14)
              at Test.bound [as _end] (/Users/astorm/Documents/github/elastic/apm-nodejs-http-client/node_modules/tape/lib/test.js:80:32)
              at Test.end (/Users/astorm/Documents/github/elastic/apm-nodejs-http-client/node_modules/tape/lib/test.js:157:10)
              at Test.bound [as end] (/Users/astorm/Documents/github/elastic/apm-nodejs-http-client/node_modules/tape/lib/test.js:80:32)
              at Server.<anonymous> (/Users/astorm/Documents/github/elastic/apm-nodejs-http-client/test/central-config.js:87:11)
              at Server.emit (events.js:315:20)
      ...
      
    //...

    not ok 1241 Error: socket hang up
      ---
        operator: error
        expected: |-
          undefined
        actual: |-
          { [Error: socket hang up] code: 'ECONNRESET' }
        at: Client.<anonymous> (/Users/astorm/Documents/github/elastic/apm-nodejs-http-client/test/central-config.js:116:11)
        stack: |-
          Error: socket hang up
              at connResetException (internal/errors.js:617:14)
              at Socket.socketCloseListener (_http_client.js:443:25)
              at Socket.emit (events.js:327:22)
              at TCP.<anonymous> (net.js:673:12)
      ...
  
We'll want/need these to pass before we can review and merge. 

If you need any help running the tests (or with anything) let us know and we'll respond as best we're able. 

Thanks again for taking the time to to make these PRs, they are appreciated.

**Comment by ddolcimascolo** [comment]
Ah yes sorry, tests were green before the rebase. I didn't pay attention. I'll fix those

**Comment by astorm** [comment]
jenkins the the tests

**Comment by astorm** [comment]
jenkins run the tests please

**Review by astorm — CHANGES_REQUESTED** [review]
@ddolcimascolo -- thanks for the merge tidying.

In reviewing this work we discovered a a pre-existing inconsistency in our current use of  http-agents in our http connections.  This isn't exactly _your_ issue but may impact what you're trying to do here. 

Currently, in a `Client`'s constructor function, we 

1. Call `this.config()`: https://github.com/elastic/apm-nodejs-http-client/blob/164775608ced60ecddfec5223505ec2927d9bbba/index.js#L59 
2. Assign a value `this._agent` _after_ the call to `this.config`: https://github.com/elastic/apm-nodejs-http-client/blob/164775608ced60ecddfec5223505ec2927d9bbba/index.js#L94

However, -- in the `config` method itself we create our request option objects (`this._conf.requestIntake` and `this._conf.requestConfig`), which uses the value of `this._agent`: https://github.com/elastic/apm-nodejs-http-client/blob/164775608ced60ecddfec5223505ec2927d9bbba/index.js#L153
    
This means there's a brief period of time between the `Client` finishing its instantiation and the next call to `.config` where the http-agent that's set at both `this._conf.requestIntake.agent` and `this._conf.requestConfig.agent` is `undefined`.  

In turn, this means that there's a single request to the configuration endpoint that will always use an _undefined_ http-agent.  Also, it _may be possible_ (depending on the length of time it takes the configuration endpoint to respond, the behavior of the elastic-agent and (_hand wave_) async callback scheduling) that the first `onStream` callback will be called before `this._conf.requestIntake.agent` is set correctly.  

Up until now this inconsistency didn't meaningfully impact the default functionality of the elastic-agent. However -- if the intent here is to ensure the elastic-agent can work with a proxy server then this may be a problem we should address now by making sure that the right http-agent is assigned before its used in any connections.

@ddolcimascolo what do you think about the above?  Do you think we need to handle this now in order to ensure things work?

**Comment by ddolcimascolo** [comment]
@astorm I totally understand and I actually spotted that while fixing the tests. I just don't have enough knowledge of the library internals, yet :)

I'm not exactly in a hurry, so yeah, I think we should handle that now in order to properly support custom agents. I'll work on it ASAP, probably by the end of the week, and update the PR.

**Comment by ddolcimascolo** [comment]
So... Now that the defined `Agent` is always used, the `_pollConfig` code does not work as expected because it expects the underlying socket to close (and emit an 'error' event) when `IncomingMessage.destroy()` is called on the received HTTP response. This is not the case when using keep alives, which is the case by default.

I've refactored the code, tests are passing but I had to remove a `client.destroy()` call, in tests, otherwise I always get a ECONNRESET error that I couldn't really explain.

Tell me what you think !

Regards,
David

**Comment by astorm** [comment]
jenkins run the tests please

**Comment by astorm** [comment]
@ddolcimascolo this is still on our radar, but we're in the middle of getting some work done for a platform release so we haven't been able to give it the thorough review it deserves.  

Taking a quick look though I had two thoughts

1. As you've discovered, the client makes _two_ HTTP requests to APM Server.  One is for the configuration polling, the second is/are the requests made that stream agent data to the request object and then on the APM Server.  Because the nature of these two HTTP requests is different (the polling is a traditional quick open/close HTTP request, the other is a "kept open for 10 seconds" request) do you think we need the ability to configure an http-agent for each one?  Or is one agent configuration enough (non-leading question -- I haven't thought hard enough about this)

2. The removal of the `client.destroy()` in the tests is a little concerning.  I'd like to understand a bit more about what the connect/disconnect flow looks like that made this necessary and whether it might impact actual agent behavior. I intend to look into this but don't have a timeframe for that -- if anyone reading wants to go deep on this please accept my pre-gratitude.

**Comment by ddolcimascolo** [comment]
Thx @astorm It's OK with me, I don't have any urging need for this one and I'd rather get it right even if that takes a bit longer.

File: index.js
Comment (astorm) [inline index.js:196]: I see here we've replaced a simple `res.destroy` call with a call to `onError`.  The old code will destroy the response object and do nothing else. The new code _does not_ destroy the response object, and also schedules a new configuration poll, and also emits an additional `request-error` event.  My naive assumption is we'd want to maintain the existing agent behavior -- can you talk a little bit about what you're trying to do here so we can figure out the best way to go about it?

File: index.js
Comment (astorm) [inline index.js:199]: I see here we've replaced a simple `res.destroy` call with a call to `onError`.  The old code will destroy the response object and do nothing else. The new code _does not_ destroy the response object, and also schedules a new configuration poll, and also emits an additional `request-error` event.  My naive assumption is we'd want to maintain the existing agent behavior -- can you talk a little bit about what you're trying to do here so we can figure out the best way to go about it?

File: test/central-config.js
Comment (astorm) [inline test/central-config.js:90]: I wasn't able to track down specifically why this line caused an error in your tests -- but it defintely seems related to losing the `destroy` calls I mentioned above.  Let's get those back in there (or talk about their intent so we can come up with a better solution) and see if this call to `destroy` is still necessary.

**Review by astorm — CHANGES_REQUESTED** [review]

File: index.js
Comment (ddolcimascolo) [reply index.js:171]: This reply is for all the comments related to that change.

This is what I was trying to explain initially, but I admit this is hard to explain... So one more try!
The old code calls `res.destroy` which will:
 - Kill the socket if keepAlives are disabled, which will fire a `error` event on the request, which will make the Client emit a `request-error` event. Yes, the `request-error` event was sent, there are tests that proves it but again, when `keepAlives` are disabled.
 - Do nothing at all when `keepAlives` are enabled, which is expected because the socket stays open for future request handling. In this case the `request-error` event logic wasn't triggered at all. That's why I'm manually calling it now
 
keepAlives are enabled by default but tests were passing because the Agent was ignored in the code, as you spotted in a previous review.

So my changes standardize the error handling, no matter if keepAlives are enabled or not, making the tests green.

I hope that my intent and the effects of the code are clearer now :)

**Comment by ddolcimascolo** [comment]
Guys, I'm going to close this one. I'll try to open a fresh PR in the future

**Pull request closed without merging**
